TranscriptService test작성

chanwooo opened the pull request:
테스트 작성했습니다.

마지막 테스트였던
void testGetRankedStudentsAsc_CourseNotExist_ThrowNoSuchCourseException_Error() {
같은경우는.. 익셉션이 발생하지않고
Optional.empty를 return하는데  익셉션을 뱉도록 모킹하였는데 이상하단 느낌이 드네요..

chanwooo committed:
TranscriptService test작성

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
+        final double averageScore = transcriptService.getAverageScore(studentID);
+
+        //then
+        Mockito.verify(studentRepository,times(1)).getStudent(studentID);
```
[minor]
- 보통 parameter를 여러 개 넣을 때는 comma(,) 이후 띄어쓰기 한 칸을 하는 것이 보기 좋습니다.
- `times(1)`은 `Mockito.verify()`의 default 이기 때문에 생략해도 괜찮습니다.
  - 이 프로젝트에서 사용하는 Mockito 버전에서도 동작이 같은지 확인 부탁드립니다.
  - 생략 가능하더라도 생략할지 안 할지 선택이 가능할 거 같은데요, @chanwooo 님 생각은 어떠신가요

chanwooo replied:
default가 times(1)이라면 생략하는게 보기 좋아보입니다. 삭제하도록하겠습니다~

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
+
+        for(var course : chanwoo.getCourses()) {
+            Mockito.verify(scoreRepository,
+                            times(1))
```
[minor]
개인적으로 불필요한 newline인 거 같습니다.

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
+        //then
+        Mockito.verify(studentRepository,times(1)).getStudent(studentID);
+
+        for(var course : chanwoo.getCourses()) {
```
[minor]
`chanwoo.getCourses().forEach(c -> ...)` 이렇게 쓸 수도 있습니다. 그러면 조금 더 modern Java 느낌이죠.

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
-        throw new UnsupportedOperationException("Not implemented yet");
+
+        final int studentID = 1;
+        final Student trey = new Student().setId(studentID).setName("Chan").setMajor("Computer Engineering")
```
[minor]
이름은 `Chan`인데 변수명은 `Trey`인 것이 조금 어색해보입니다.

f-lab-trey replied:
복붙하기가 귀찮고 지저분해보이면 `Student`를 생성하는 Factory method를 하나 만들어도 좋을 거 같습니다.

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
+
+        Mockito.when(studentRepository.getStudent(studentID))
+                .thenReturn(Optional.of(trey));
+
```
`ScoreRepository`에 대한 Stubbing을 생략하지말고, `ScoreRepository.getScore()` 했을 때 `Optional.empty()`를 리턴하는 Stubbing을 여기 넣는 게 테스트 내용을 더 잘 반영하는 거라고 생각합니다.

추가질문) Stubbing을 하지 않으면 테스트 코드가 실행되었을 때 어떤 Exception이 발생하나요?

chanwooo replied:
1. 그렇네요 서비스 코드안에서 getScore를 사용하는데도 Stubbing하지않고 지나쳐버렸네요..
아래와같이 Stubbing하도록 하였습니다.
```
        Mockito.when(scoreRepository.getScore(studentID, courseID))
                .thenReturn(Optional.empty());
```

2. NoSuchStudentException이 발생합니다.

f-lab-trey replied:
하고 싶었던 질문은 왜 `NoSuchStudentException`이 발생할까요 인데 제가 정확하지 못했네요. Stubbing 하지 않은 method를 호출했을 때 코드가 어떻게 동작하는지 알고 있는가에 대한 질문입니다.

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
+
+        Assertions.assertThrows(NoSuchScoreException.class, () -> transcriptService.getAverageScore(studentID));
+
+
```
[minor]
불필요한 empty line 입니다.

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
+                .thenReturn(Optional.of(KOREAN));
+
+        Mockito.when(scoreRepository.getScores(KOREAN.getId()))
+                .thenReturn(Map.of(1,new Score().setScore(100),2,new Score().setScore(90)));
```
[minor]
ditto (comma 다음 띄어쓰기)

chanwooo replied:
indent를 이런식으로 해봤어요. 제가 보기엔 훨씬 깔끔한거같은데 멘토님 보기에 어떠신가요?
```
        Mockito.when(scoreRepository.getScores(KOREAN.getId()))
                .thenReturn(Map.of(
                                1, new Score().setScore(100),
                                2, new Score().setScore(90)
                        )
                );
```

f-lab-trey replied:
넵, 제가 보기에도 이쪽이 더 좋아보입니다. 👍🏼

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
+        List<Student> students = transcriptService.getRankedStudentsAsc(KOREAN.getId());
+
+        //then
+        Assertions.assertEquals(chan1.getId(),students.get(0).getId());
```
[q]
첫 번째 항목만 체크해도 괜찮을까요?

chanwooo replied:
부족한것같네요. 차라리 순서대로 만든 리스트와 비교하는게 좋을거같아요.
아래와같이 비교하는 라인을 추가했습니다.
```
        Assertions.assertEquals(List.of(chan1, chan2), students);

```

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
 
     @Test
     @DisplayName("getRankedStudentAsc()를 호출하면, 입력으로 주어진 course를 수강하는 모든 학생들의 리스트를 성적의 내림차순으로 리턴한다.")
     void testGetRankedStudentsAsc_HappyCase_VerifyReturnedValueAndInteractions_Success() {
```
[q]
이 테스트로 `TranscriptServiceImple.getRankedStudentsAsc()`의 HappyCase를 모두 커버했다고 할 수 있을까요?

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
+                .thenReturn(List.of(chan1, chan2));
+
+        //when
+        List<Student> students = transcriptService.getRankedStudentsAsc(KOREAN.getId());
```
[minor]
귀찮지만 `final`을 모두 붙이는 것이 바람직합니다.

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -80,31 +85,90 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
     void testGetRankedStudentsAsc_CourseNotExist_ThrowNoSuchCourseException_Error() {
-        // TODO:
-        throw new UnsupportedOperationException("Not implemented yet");
+        Mockito.when(courseRepository.getCourse(1))
+                .thenThrow(NoSuchCourseException.class);
```
`Optional.empty()`를 이용해서 고쳐봅시다.

chanwooo replied:
아.. 지금 다시보니까 설명만 보고 Service에 대해 콜을 하나도 안하고 repository에 대해서만 호출을 했네요.
아래와같이 수정했습니다~

```
    Mockito.when(courseRepository.getCourse(courseID))
                .thenReturn(Optional.empty());

        Assertions.assertThrows(NoSuchCourseException.class, () -> transcriptService.getRankedStudentsAsc(courseID));
```

chanwooo committed:
코드리뷰 반영

공통
- Student 생성하는 팩토리 메서드 생성

testGetAverageScore_HappyCase_VerifyNumberOfInteractions_Success
- for loop 대신 .forEach() 사용
- 변수 이름 수정 trey -> chan
- Mockito verify times(1) 제거 (times(1)이 defalut)

testGetAverageScore_ScoreNotExist_ThrowNoSuchScoreException_Error
- scoreRepository.getScore에 대해 Stubbing 하는 구문 추가

testGetRankedStudentsAsc_HappyCase_VerifyReturnedValueAndInteractions_Success
- Map.Of 내용 Indent 변경
- List.of로 만든 List와 getRankedStudnetAsc의 결과 비교하는 Assertion 추가

testGetRankedStudentsAsc_CourseNotExist_ThrowNoSuchCourseException_Error
- Stubbing, Assert 수정

f-lab-trey commented on src/test/java/com/flab/service/TranscriptServiceImplTest.java:
```diff
@@ -106,69 +105,78 @@ void testGetAverageScore_HappyCase_VerifyNumberOfInteractions_Success() {
 
     }
+
+    Student createStudent(int id) {
```
[minor]
이 method는 다른 데서 쓸 일이 없기 때문에 `private`으로 하는 게 조금 더 적절하다고 생각합니다.

f-lab-trey reviewed (approved):
LGTM

f-lab-trey commented:
merge 권한이 있는지는 잘 모르겠는데, 혹시 있더라도 merge 하지 말아주세요!

chanwooo committed:
createStudent 접근지시자 변경 default -> private
